### PR TITLE
Tuto ddpm

### PR DESCRIPTION
**Pull request by olivier-bernard-creatis**
The tuto is complete, don't spend too much time to review the equations ;)

### CODE REVIEW

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 1]
```diff
@@ -0,0 +1,786 @@
...
+
+> The use of a time-averaging smoothed version of $$\mathcal{L}_{VLB}$$ was key to stabelize the optimization of $$\mathcal{L}_{VLB}$$ due to the presence of noisy gradients coming from this term.
+
+&nbsp;
+
+### Improve sampling speed
+
+- It is very slow to generate a sample from DDPM by following the Markov chain of the reverse process as $$T$$ can be up to one or a few thousand steps
+
+- Producing a single sample can takes several minutes on a modern GPU.
```
Comment (jodoin) [inline collections/_tutorials/2023-11-30-tutorial-ddpm.md:697]: takes

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 2]
```diff
@@ -0,0 +1,786 @@
...
+- One simple way to improve this is to run a strided sampling scheduler.
+
+- For a model trained with $$T$$ diffusion steps, the sampling procedure only use $$[T/S]$$ steps to reduce the process from $$T$$ to $$S$$ steps. 
+
+- The new sampling schedule for data generation is $$\{\tau_1,\cdots,\tau_S\}$$ where $$\tau_1 < \tau_2 < \cdots < \tau_S \in [1,T]$$ and $$S<T$$.
+
+
+&nbsp;
+
+- Another algorithm, DDIM named [Denoising Diffusion Implicit Model](https://arxiv.org/abs/2010.02502), modified the forward diffusion process making it non-Markovian to speed up the generation. The key is to change the forward diffusion into a non-Markovian process while keeeping the reverse process Markovian but with fewer steps.
```
Comment (jodoin) [inline collections/_tutorials/2023-11-30-tutorial-ddpm.md:708]: modifies

Comment (jodoin) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:708]: keeping

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 3]
```diff
@@ -0,0 +1,786 @@
...
+
+- The full forward process can be modeled as: 
+
+$$q(x_{1:T} \mid x_0) = \prod_{t=1}^{T}{q(x_t \mid x_{t-1})}$$
+
+- Based on the definition of the forward process, the conditional distribution $$q(x_t \mid x_{t-1})$$ can be efficiently represented as:
+
+<div style="text-align:center">
+<span style="color:#00478F">
+$$q(x_t \mid x_{t-1}) = \mathcal{N}\left((\sqrt{1 - \beta _t}) \, x_{t-1},\beta _t \, \textbf{I}\right)$$
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:198]: This is a good point. This is the modeling proposed in the seminal paper, might be to introduced some non linearity for the mean value during the diffusion process

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 4]
```diff
@@ -0,0 +1,786 @@
...
+& \text{if} \quad \beta_t=1 , \quad \text{then} \quad q(x_t \mid x_{t-1})=\mathcal{N}(0,\textbf{I})
+\end{align}$$
+
+&nbsp;
+
+- A nice property of the forward process is that we can sample $$x_t$$ at any arbitrary time step $$t$$ in a closed form using [reparametrization trick](#reparameterization-trick) and the property of the [sum of normally distributed variables](#sum-of-normally-distributed-variables), as shown below:
+
+$$q(x_t \mid x_{t-1}) = \mathcal{N}\left((\sqrt{1 - \beta _t}) \, x_{t-1},\beta _t \, \textbf{I}\right)$$
+
+$$x_t = (\sqrt{1 - \beta_t}) \, x_{t-1} + \sqrt{\beta_t} \, \epsilon_{t-1}$$
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:223]: Yes, exactly

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 5]
```diff
@@ -0,0 +1,786 @@
...
+
+- If we are able to reverse the diffusion process from $$q(x_{t-1} \mid x_t)$$, we will be able to generate a sample $$x_0$$ from a Gaussian noise input $$x_T \sim \mathcal{N}(0,\mathbf{I})$$ !
+
+- Unfortunately, using the [Bayes theorem](#bayes-theorem) and keeping in mind that $$q(x_t)$$ is an unknown, one can easy see that $$q(x_{t-1} \mid x_t)$$ is intractable.
+
+$$q(x_{t-1} \mid x_t) = \frac{q(x_{t} \mid x_{t-1}) \, q(x_{t-1})}{q(x_{t})}$$
+
+&nbsp;
+
+> It is noteworthy that the reverse conditional probability is tractable when conditioned on $$x_0$$. Indeed, thanks to Bayes theorem $$\, q(x_{t-1} \mid x_t, x_0)$$ = $$\frac{q(x_t \mid x_{t-1}, x_0)q(x_{t-1} \mid x_0)}{q(x_t \mid x_0)}$$=$$\frac{q(x_t \mid x_{t-1})q(x_{t-1} \mid x_0)}{q(x_t \mid x_0)}$$, where all distributions are known from the forward process.
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:310]: no, markov chain implies that the state at time t only depends on the previous step, so we have q(xt|xt-1,x0) = q(xt|xt-1). However you can not say that q(xt|x0) = q(xt), you need to know the input image x0 to know noisy image xt ;)

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 6]
```diff
@@ -0,0 +1,786 @@
...
+& \tilde{\mu}_t(x_t,x_0) = \frac{1}{\sqrt{\alpha_t}} (x_t - \frac{1-\alpha _t}{\sqrt{1- \bar \alpha _t}} \epsilon _t)\\
+& \tilde{\beta}_t = \frac{1 - \bar{\alpha}_{t-1}}{1 - \bar{\alpha}_t} \cdot \beta_t \\
+\end{align}$$
+</span>
+</div>
+
+
+&nbsp;
+
+- However, if $$\beta_t$$ is small enough, $$q(x_{t-1} \mid x_t)$$ will also be Gaussian.
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:336]: No, we're talking about q(xt−1∣xt), not q(xt−1∣xt, x0).

q(xt−1∣xt, x0) is Gaussian, it is demonstrated just above. 

the hypothesis that q(xt−1∣xt) is Gaussian will be used after for the approximated distribution p_theta(xt-1|xt)

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 7]
```diff
@@ -0,0 +1,786 @@
...
+
+& = \mathbb{E}_{x_{0:T} \sim q}\left[-\log\left(p_{\theta}(x_{T})\right) + \sum^T_{t=2} \log\left( \frac{ q(x_t \vert x_{t-1})}{p_{\theta}(x_{t-1} \vert x_t)}\right) + \log\left( \frac{ q(x_1 \vert x_0)}{p_{\theta}(x_0 \vert x_1)}\right)\right] 
+\end{align} $$
+
+&nbsp;
+
+- Using the [markov chain property](#markov-chain) with Bayes, the above expression can be reformulated as:
+
+$$ \begin{align} 
+\mathcal{L}_{VLB}  & = \mathbb{E}_{x_{0:T} \sim q} \left[-\log \left(p_{\theta}(x_{T}) \right) + \sum^T_{t=2} \log \left( \frac{ q(x_{t-1} \vert x_t,x_0)q(x_t \mid x_0)}{p_{\theta}(x_{t-1} \vert x_t)q(x_{t-1} \mid x_0)} \right) + \log \left( \frac{ q(x_1 \mid x_0)}{p_{\theta}(x_0 \mid x_1)} \right) \right] \\
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:424]: The demonstration is given in section "Markov chain properties" / "Baye theorem" at the begining of the tuto

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 8]
```diff
@@ -0,0 +1,786 @@
...
+& \mathcal{L}_0 = -\log \left( p_{\theta}(x_0 \mid x_1)\right)\\
+\end{align}$$
+</span>
+</div>
+
+&nbsp;
+
+**$$\mathcal{L}_T:$$ Constant Term**
+
+- $$\mathcal{L}_T$$ is a constant matrix since $$x_0$$ is known, the forward process is deterministic, $$q(x_T \mid x_0)$$ is a constant and $$p_\theta(x_{T})$$ is a sample from a known Gaussian distribution.
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:462]: I agree that the sentence is not clear. I rewrited it using the justification given in the seminal article ;)

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 9]
```diff
@@ -0,0 +1,786 @@
...
+
+&nbsp;
+
+> It is thus possible to generate a random image $$\,x_t$$ of the input image $$\,x_0$$ at any time $$\,t$$ of the diffusion process with a known noise $$\,\epsilon_t \in \mathcal{N}(0,\mathbf{I})$$. 
+
+&nbsp;
+
+- In diffusion model, the deep learning architecture aims to estimate $$\epsilon_t$$ from $$x_t$$ at time $$t$$. The estimated noise is denoted $$\epsilon_{\theta}(x_t,t)$$.
+
+- Recall that $$p_{\theta}(x_{t-1} \vert x_t) = \mathcal{N}(\mu_{\theta}(x_t,t), \Sigma_{\theta}(x_t,t))$$ with $$\mu_{\theta}(x_t,t) = \frac{1}{\sqrt{\alpha_t}} \left(x_t - \frac{1-\alpha_t}{\sqrt{1- \bar{\alpha}_t}}\right) \epsilon_{\theta}(x_t,t)$$.
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:617]: Bien vu, merci !

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jodoin) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:310]: Really?!  i just asked to a statistitian and he doesn't agree with this assertion.

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jodoin) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:422]: Yeah... this comes down to my previous concern about the dependancy of x_t and x_t+1 to x_0

File: collections/_tutorials/2023-11-16-tutorial-ddpm.md [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+x + \frac{1}{255} & \text{if $x < 1$}
+\end{cases}
+$$
+$$\delta _{-} (x) = \begin{cases}
+-\infty & \text{if $x = -1$} \\
+x - \frac{1}{255} & \text{if $x > -1$}
+\end{cases}
+$$
+
+where $$D$$ is the data dimensionality and $$i$$ indicate the extraction of one coordinate.
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-16-tutorial-ddpm.md]: Sorry, I missed this comment, i just have correct it.

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 11]
```diff
@@ -0,0 +1,786 @@
...
+
+&nbsp;
+
+**$$\mathcal{L}_t:$$ Stepwise denoising terms**
+
+- Recall that $$p_{\theta}(x_{t-1} \vert x_t) = \mathcal{N}(\mu_{\theta}(x_t,t), \Sigma_{\theta}(x_t,t))$$ and $$q(x_{t-1} \vert x_t, x_0) = \mathcal{N}(\tilde{\mu}_t(x_t,x_0), \tilde{\beta}_t \cdot \textbf{I})$$, <spam style="color:#00478F">the idea is first to focus on the mean terms</spam> and train a neural network $$\mu_{\theta}$$ to predict $$\tilde{\mu}_t = \frac{1}{\sqrt{\alpha_t}} (x_t - \frac{1-\alpha_t}{\sqrt{1- \bar{\alpha}_t}} \epsilon _t)$$
+
+&nbsp;
+
+> Because $$x_t$$ is available as input at training time, we can reparameterize the Gaussian noise term instead to make it predict $$\epsilon_t$$ from the input $$x_t$$ at time step
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:496]: Sorry, I missed this comment, i just have correct it.

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:310]: so we have a problem ;)

[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:310]: We will check from our side to solve this math issue

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md:422]: yep, we check this from our side and come back to you asap

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 12]
```diff
@@ -0,0 +1,786 @@
...
+
+&nbsp;
+
+<div style="text-align:center">
+<span style="color:#00478F">
+$$\mu_{\theta}(x_t,t) = \frac{1}{\sqrt{\alpha_t}} \left(x_t - \frac{1-\alpha_t}{\sqrt{1- \bar{\alpha}_t}}\right) \epsilon_{\theta}(x_t,t)$$
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md]: I effectively saw yesterday that I forget to fix ;)
Now this is done !

File: collections/_tutorials/2023-11-30-tutorial-ddpm.md [hunk 13]
```diff
@@ -0,0 +1,786 @@
...
+-\infty & \text{if $x = -1$} \\
+x - \frac{1}{255} & \text{if $x > -1$}
+\end{cases}
+$$
+
+- An independant discrete decoder is set to obtain the corresponding log likelihood.
+
+&nbsp;
+
+**$$\mathcal{L}_t:$$ Stepwise denoising terms**
```
[the remark this answers is not in the document]
Comment (olivier-bernard-creatis) [reply collections/_tutorials/2023-11-30-tutorial-ddpm.md]: you're right, thanks

**Review by nathanpainchaud — APPROVED** [review]

**Pull request merged**